Reader: Configure `ReaderPostCardCell` with a `ReaderPost`

wargcm opened the pull request:
Part of #21645 
Merges into #21685

## Description

Configures the new reader card post UI with reader posts.

> **Note**
> The following will be in future PR's:
> - Button actions
> - Safe area fixes
> - Accessibility fixes

## Screenshots

| Light | Dark |
|------|------|
| ![light](https://github.com/wordpress-mobile/WordPress-iOS/assets/2454408/83d7389f-3775-4ca4-bcbc-3936b7dc7bd1) | ![dark](https://github.com/wordpress-mobile/WordPress-iOS/assets/2454408/a2ddd319-a466-48b4-9258-f445aaca704d) |

## Testing

To test:
- Launch Jetpack and login
- Enable the `Reader Improvements v1` feature flag if necessary
- Tap on Reader
- 🔎 **Verify**:
  - Cells are configured with reader post data
  - The appropriate controls appear depending on the blog
- Tap on `Likes` at the top
- 🔎 **Verify** liked posts show the correct button and image
- Tap on `Automattic`
- 🔎 **Verify**:
  - The avatar view appears with the author's image
  - The title has the author's name
  - The reblog button does not appear

## Regression Notes
1. Potential unintended areas of impact
N/A

2. What I did to test those areas of impact (or what existing automated tests I relied on)
N/A

3. What automated tests I added (or what prevented me from doing so)
N/A

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding unit tests for my changes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

UI Changes testing checklist:
- [ ] Portrait and landscape orientations.
- [ ] Light and dark modes.
- [ ] Fonts: Larger, smaller and bold text.
- [ ] High contrast.
- [ ] VoiceOver.
- [ ] Languages with large words or with letters/accents not frequently used in English.
- [ ] Right-to-left languages. (Even if translation isn’t complete, formatting should still respect the right-to-left layout)
- [ ] iPhone and iPad. 
- [ ] Multi-tasking: Split view and Slide over. (iPad)

wargcm committed:
Configure `ReaderCardView` with a `ReaderPost`

wargcm commented on WordPress/Classes/Utility/Date+Formats.swift:
```diff
@@ -0,0 +1,39 @@
+import Foundation
+
+extension Date {
+    /// Extracts the time from the passed date
+    func toLocalTime() -> String {
+        let formatter = DateFormatter()
+        formatter.timeStyle = .short
+        return formatter.string(from: self)
+    }
+
+    func toLocal24HTime() -> String {
+        let formatter = DateFormatter()
+        formatter.dateFormat = "HH:mm"
+        return formatter.string(from: self)
+    }
```
These were existing methods from `Date+TimeStrings.swift`.

wargcm committed:
Use a view model for `ReaderPostCardCell`

dvdchr commented on WordPress/Classes/Utility/Date+Formats.swift:
```diff
@@ -0,0 +1,39 @@
+    /// - Example: 2 days ago
+    /// - Example: Jan 22, 2017
+    ///
+    func toShortString() -> String {
```
Ah, I'm using `toMediumString()` from `WordPressShared` which doesn't abbreviate the units (basically keeping the previous implementation). 

We should use the same date format in both the cards and the detail screen, and I noticed in p1696262754792909-slack-C05N140C8H5 Android also uses the abbreviated format. I can update the date presentation on my end later to this one.

dvdchr commented on WordPress/Classes/ViewRelated/Reader/ReaderPostCardCellViewModel.swift:
```diff
@@ -0,0 +1,149 @@
+        contentProvider.titleForDisplay()
+    }
+
+    private var featuredImageIdealSize: CGSize? {
+        guard let window = WordPressAppDelegate.shared?.window else {
+            return nil
+        }
+
+        let windowWidth = window.screen.bounds.width
+        let safeAreaOffset = window.safeAreaInsets.left + window.safeAreaInsets.right
+        let width = windowWidth - safeAreaOffset - ReaderPostCardCell.Constants.ContentStackView.margins * 2
+        let height = width * ReaderPostCardCell.Constants.FeaturedImage.heightAspectMultiplier
+        return CGSize(width: width, height: height)
+    }
```
Some images appear stretched while testing. Is this what you meant by "Safe area fixes" not being included in this PR? Here's an example:

<img src="https://github.com/wordpress-mobile/WordPress-iOS/assets/1299411/77492bb5-6eb1-43c8-b7c7-26620f8ff6f6" width=400>

wargcm replied:
The safe area fixes are something different.

I wasn't really sure about which content mode to use for the image view.  We could use `.scaleAspectFit` but that can cause a lot of unused space in the image view. And `.scaleAspectFill` causes content to be cut off. Some examples:

| Default | `.scaleAspectFit` | `.scaleAspectFill` |
|---------|------------------|-------------------|
| ![Screenshot 2023-10-06 at 5 24 40 PM](https://github.com/wordpress-mobile/WordPress-iOS/assets/2454408/5d39f716-d356-4078-9198-fd25948633da) | ![Screenshot 2023-10-06 at 5 25 44 PM](https://github.com/wordpress-mobile/WordPress-iOS/assets/2454408/0d2b7e13-f90b-4269-bde8-18e2afdc8d51) | ![Screenshot 2023-10-06 at 5 27 43 PM](https://github.com/wordpress-mobile/WordPress-iOS/assets/2454408/19b41a5b-0e8b-480c-a567-aac33c95dc7f) |

Current card:

![Screenshot 2023-10-06 at 5 36 21 PM](https://github.com/wordpress-mobile/WordPress-iOS/assets/2454408/b67132d5-592e-4c76-9140-586cd65a28da)

@osullivanchris Do you have any particular preference for which one we should use?

osullivanchris replied:
I think we should use .scaleAspectFill which appears to keep the original content aspect ratio, but frame it. 

We discussed how to define the image sizes on Android. To solve for tall portrait images, what I came up with was to have a sensible max-height e.g. 500pt, where the card won't be taller than the visible content area. For super wide images, we also thought we might need a rule. Here it seemed like a min-height (e.g. 120pt) would prevent a situation where the original image was really wide, and renders as 20px tall on our stream.

In both cases I'd have that max height / min-width as a frame on the image. And the image would be centred horizontally and vertically within. 

Maybe later we could have some Ai which identifies the most important area of the image and automatically centres / zooms :)

wargcm replied:
I updated the image's content mode in: https://github.com/wordpress-mobile/WordPress-iOS/pull/21687/commits/0ab03d5e17554402b33ffd3e73798917feeab44d.

Did we want the images to be a dynamic height? I purposely made all the images the same width/height, so we wouldn't have differently sized cards.

dvdchr commented on WordPress/Classes/ViewRelated/Reader/ReaderPostCardCellViewModel.swift:
```diff
@@ -0,0 +1,149 @@
+    var siteTitle: String? {
+        if let post = contentProvider as? ReaderPost, post.isP2Type(), let author = contentProvider.authorForDisplay() {
+            let strings = [author, contentProvider.blogNameForDisplay?()].compactMap { $0 }
+            return strings.joined(separator: " ▸ ")
```
This will become an issue in RTL since you need it to be reversed (even the arrow):

```
Site name ◂ Author name
```

wargcm replied:
Fixed with https://github.com/wordpress-mobile/WordPress-iOS/pull/21687/commits/1d194613f22a5ad67508f3521ac982256cf19cad. Thanks for the catch!

Example:
![rtl](https://github.com/wordpress-mobile/WordPress-iOS/assets/2454408/e87d9391-6641-4832-8192-a2d58e6f3e72)

dvdchr commented on WordPress/Classes/ViewRelated/Reader/ReaderPostCardCellViewModel.swift:
```diff
@@ -0,0 +1,149 @@
+        guard let dateForDisplay = contentProvider.dateForDisplay()?.toShortString() else {
+            return nil
+        }
+        return siteTitle != nil ? "• \(dateForDisplay)" : dateForDisplay
```
Same RTL issue as the above, but in this case, you probably just need to place the dot at the end on RTL layout.

wargcm replied:
Fixed with https://github.com/wordpress-mobile/WordPress-iOS/pull/21687/commits/1d194613f22a5ad67508f3521ac982256cf19cad. Thanks for the catch!

Example:
![rtl](https://github.com/wordpress-mobile/WordPress-iOS/assets/2454408/e87d9391-6641-4832-8192-a2d58e6f3e72)

dvdchr reviewed (approved):
LGTM. Found some accessibility issues, but I'll let you decide whether to address them here or in the later PRs!

osullivanchris commented:
This PR is beautiful. Thanks for making things so faithful to the design. It looks/feels way better. I have a couple of tweaks. 

First two specs I was on the fence about, and have changed my mind interacting with the real thing:
- let's change the site avatar to 20pt rather than 24pt. It will be more similar in size to the site name and tighten up the vertical spacing on the card.
- Let's reduce the horizontal padding between the actions to 16pt rather than 24pt. They look too far apart. And it's easy to tap the right one.

A few more minor issues:
- I don't think I had covered it but I think the "Liked" icon and label should be Jetpack Green 50 rather than remaining grey.
- Is it possible to say '4d ago' rather than '4 days ago'. 'days' is a long string compared with 'hr' when we're doing these contextual dates. I know that some of this comes with the system, so don't change it if its hard.
- If possible can we have the timestamp at 13pt regular footnote, while keeping the site name at subheading 15pt semibold. This is what was in the design, but easily missed.
- Cross-posts in my Automattic stream still have the serif font for their titles.

Generally again looks great and look forward to seeing more.

wargcm committed:
Merge branch 'issue/21645-reader-card-2' into issue/21645-reader-card-3

wargcm committed:
Fix RTL layout for post date and author title

wargcm committed:
Set featured image's content to `.scaleAspectFill`

wargcm committed:
Reduce icon sizes to `20.0` and set content mode

wargcm committed:
Reduce button spacing to `16.0`

wargcm committed:
Update 'Liked' button color to Jetpack green

wargcm commented:
@osullivanchris 

> let's change the site avatar to 20pt rather than 24pt. It will be more similar in size to the site name and tighten up the vertical spacing on the card.

Added in https://github.com/wordpress-mobile/WordPress-iOS/pull/21687/commits/73b1d7bd73fef07799720b9c2fe30734b254e465. I still need to add the borders to these icons. I should have that PR today, and I'll add you as a reviewer to it too.

> Let's reduce the horizontal padding between the actions to 16pt rather than 24pt. They look too far apart. And it's easy to tap the right one.

Added with https://github.com/wordpress-mobile/WordPress-iOS/pull/21687/commits/6f087aeddca0852b088a8668adfe7da926e214e3.

> I don't think I had covered it but I think the "Liked" icon and label should be Jetpack Green 50 rather than remaining grey.

Sounds good, I updated it to use Jetpack green with https://github.com/wordpress-mobile/WordPress-iOS/pull/21687/commits/1dd90e14f6aa491ec8afccd5a6cbb9fadafbea24.

> Is it possible to say '4d ago' rather than '4 days ago'. 'days' is a long string compared with 'hr' when we're doing these contextual dates. I know that some of this comes with the system, so don't change it if its hard.

Unfortunately, this is controlled by a system formatter. I tried picking the shortest one available, ~but it still uses 'days' instead of abbreviating it~.

Edit: It looks like the formatter I used changes based on the system version installed. In those pictures, it's using iOS 16.4 but in iOS 17, it abbreviates the days to 'd'.

> If possible can we have the timestamp at 13pt regular footnote, while keeping the site name at subheading 15pt semibold. This is what was in the design, but easily missed.

Yep, this should be fixed now! David caught that in the last PR and I merged those fixes into this one.

> Cross-posts in my Automattic stream still have the serif font for their titles.

The cross-posts are a separate card that we still need to address. Those changes will be done in a future PR.

Thanks for reviewing this!

osullivanchris commented:
@wargcm thanks for all those quick fixes. All looking much better now 👍 

No problem re; the day string. Understood. 

And happy to look at x-posts and borders in separate PRs.
